File: packages/zod/src/share.ts
import { z } from 'zod'
import { ZModel } from './utils.js'

export const ZShareVisibility = z.enum(['public', 'unlisted'])
export const ZShareStatus = z.enum(['active', 'disabled', 'removed'])
export const ZBorrowStatus = z.enum(['active', 'returned', 'expired', 'revoked'])
export const ZReportReason = z.enum(['copyright', 'abuse', 'spam', 'other'])
export const ZReportStatus = z.enum(['open', 'in_review', 'resolved', 'rejected'])

export const ZShare =
	z
		.object({
			ebookId: z.string().uuid(),
			ownerUserId: z.string().uuid(),
			titleOverride: z.string().optional(),
			description: z.string().optional(),
			visibility: ZShareVisibility,
			status: ZShareStatus,
			borrowDurationHours: z.number().int().positive(),
			maxConcurrentBorrows: z.number().int().positive(),
		})
		.extend(ZModel.shape)

export const ZStoreShareDTO = z.object({
	ebookId: z.string().uuid(),
	titleOverride: z.string().max(255).optional(),
	description: z.string().optional(),
	visibility: ZShareVisibility.optional(),
	status: ZShareStatus.optional(),
	borrowDurationHours: z.number().int().positive(),
	maxConcurrentBorrows: z.number().int().positive().optional(),
})

export const ZUpdateShareDTO = z.object({
	titleOverride: z.string().max(255).optional(),
	description: z.string().optional(),
	visibility: ZShareVisibility.optional(),
	status: ZShareStatus.optional(),
	borrowDurationHours: z.number().int().positive().optional(),
	maxConcurrentBorrows: z.number().int().positive().optional(),
})

export const ZBorrow = z.object({
	id: z.string().uuid(),
	createdAt: z.string().datetime(),
	updatedAt: z.string().datetime(),
	shareId: z.string().uuid(),
	borrowerUserId: z.string().uuid(),
	startedAt: z.string().datetime(),
	dueAt: z.string().datetime(),
	returnedAt: z.string().datetime().optional(),
	expiredAt: z.string().datetime().optional(),
	status: ZBorrowStatus,
	legalAcknowledgedAt: z.string().datetime(),
})

export const ZBorrowShareDTO = z.object({
	legalAcknowledged: z.literal(true),
})

export const ZShareReview =
	z
		.object({
			shareId: z.string().uuid(),
			userId: z.string().uuid(),
			rating: z.number().int().min(1).max(5),
			reviewText: z.string().optional(),
		})
		.extend(ZModel.shape)

export const ZUpsertShareReviewDTO = z.object({
	rating: z.number().int().min(1).max(5),
	reviewText: z.string().optional(),
})

export const ZShareReport = z.object({
	id: z.string().uuid(),
	createdAt: z.string().datetime(),
	updatedAt: z.string().datetime(),
	shareId: z.string().uuid(),
	reporterUserId: z.string().uuid(),
	reason: ZReportReason,
	details: z.string().optional(),
	status: ZReportStatus,
	reviewedByUserId: z.string().uuid().optional(),
	reviewedAt: z.string().datetime().optional(),
	resolutionNote: z.string().optional(),
})

export const ZCreateShareReportDTO = z.object({
	reason: ZReportReason,
	details: z.string().optional(),
})

File: apps/api/internal/interface/http/dto/share.go
package dto

import (
	"github.com/go-playground/validator/v10"
	"github.com/google/uuid"
	applicationdto "github.com/jeheskielSunloy77/libra-link/internal/application/dto"
	"github.com/jeheskielSunloy77/libra-link/internal/domain"
)

type StoreShareRequest struct {
	EbookID              uuid.UUID              `json:"ebookId" validate:"required"`
	TitleOverride        *string                `json:"titleOverride" validate:"omitempty,max=255"`
	Description          *string                `json:"description"`
	Visibility           domain.ShareVisibility `json:"visibility" validate:"omitempty,oneof=public unlisted"`
	Status               domain.ShareStatus     `json:"status" validate:"omitempty,oneof=active disabled removed"`
	BorrowDurationHours  int                    `json:"borrowDurationHours" validate:"required,gt=0"`
	MaxConcurrentBorrows int                    `json:"maxConcurrentBorrows" validate:"omitempty,gte=1"`
}

func (d *StoreShareRequest) Validate() error {
	return validator.New().Struct(d)
}

func (d *StoreShareRequest) ToUsecase() *applicationdto.StoreShareInput {
	maxConcurrent := d.MaxConcurrentBorrows
	if maxConcurrent == 0 {
		maxConcurrent = 1
	}

	return &applicationdto.StoreShareInput{
		EbookID:              d.EbookID,
		TitleOverride:        d.TitleOverride,
		Description:          d.Description,
		Visibility:           d.Visibility,
		Status:               d.Status,
		BorrowDurationHours:  d.BorrowDurationHours,
		MaxConcurrentBorrows: maxConcurrent,
	}
}

type UpdateShareRequest struct {
	TitleOverride        *string                 `json:"titleOverride" validate:"omitempty,max=255"`
	Description          *string                 `json:"description"`
	Visibility           *domain.ShareVisibility `json:"visibility" validate:"omitempty,oneof=public unlisted"`
	Status               *domain.ShareStatus     `json:"status" validate:"omitempty,oneof=active disabled removed"`
	BorrowDurationHours  *int                    `json:"borrowDurationHours" validate:"omitempty,gt=0"`
	MaxConcurrentBorrows *int                    `json:"maxConcurrentBorrows" validate:"omitempty,gte=1"`
}

func (d *UpdateShareRequest) Validate() error {
	return validator.New().Struct(d)
}

func (d *UpdateShareRequest) ToUsecase() *applicationdto.UpdateShareInput {
	return &applicationdto.UpdateShareInput{
		TitleOverride:        d.TitleOverride,
		Description:          d.Description,
		Visibility:           d.Visibility,
		Status:               d.Status,
		BorrowDurationHours:  d.BorrowDurationHours,
		MaxConcurrentBorrows: d.MaxConcurrentBorrows,
	}
}

type BorrowShareRequest struct {
	LegalAcknowledged bool `json:"legalAcknowledged" validate:"required,eq=true"`
}

func (d *BorrowShareRequest) Validate() error {
	return validator.New().Struct(d)
}

type UpsertShareReviewRequest struct {
	Rating     int16   `json:"rating" validate:"required,min=1,max=5"`
	ReviewText *string `json:"reviewText"`
}

func (d *UpsertShareReviewRequest) Validate() error {
	return validator.New().Struct(d)
}

type CreateShareReportRequest struct {
	Reason  domain.ReportReason `json:"reason" validate:"required,oneof=copyright abuse spam other"`
	Details *string             `json:"details"`
}

func (d *CreateShareReportRequest) Validate() error {
	return validator.New().Struct(d)
}
